syntax = "proto3";

message RobotStatus {
  bool is_active = 1;
  int32 robot_id = 2;
}

message RobotRegistration {
  int32 robot_id = 1;
  string robot_name = 2;  
  string robot_role = 3;
}

message HeartbeatRequest {
  int32 robot_id = 1;
}

message HeartbeatResponse {
  bool success = 1;
  bool not_registered = 2;
}

message ServerRequest {
  string name = 1;
}

message DBResponse{
  string captain_name = 1;
  int32 captain_id = 2;
  repeated string player_name = 3;
  repeated int32 player_id = 4;
  int32 highest_id = 5;
}

message Ping{
  bool is_active = 1;
}

message Request {
  bool ans = 1;
}

message Response {
  int32 robot_id = 1;
  string robot_role = 2;
  bool start = 3;
}

message NewCaptainRequest {
  int32 robot_id = 1;
}

message NewCaptainResponse {

}

service RobotControl {
  rpc RegisterRobot(RobotRegistration) returns (RobotStatus);
  rpc GetRobotStatus(RobotRegistration) returns (RobotStatus);
  rpc SendHeartbeat(HeartbeatRequest) returns (HeartbeatResponse);
  rpc SyncDB(ServerRequest) returns (DBResponse);
  rpc Ping_gRPC(ServerRequest) returns (Ping);
  rpc StartElection(stream Request) returns (stream Response);
  rpc ResultElection(NewCaptainRequest) returns (NewCaptainResponse);
}


